Ob lesim

drghirlanda opened the pull request:
### Brief summary of what the package does

This package integrates org-mode and lesim-mode (a major mode for Learning Simulator files, see learningsimulator.org)

### Direct link to the package repository

https://github.com/drghirlanda/ob-lesim

### Your association with the package

Author and maintainer

### Relevant communications with the upstream package maintainer

Linting and byte-compiling issue a few warnings about unused variables and unorthodox function names. However, both of these are mandated by the org-babel interface.

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
This branch is also including your earlier lesim-mode commits; this won't be a problem, but for future submissions you should make sure your fork-point is up to date.

And, thank you; here's a quick first pass --

`ob-lesim.el` with byte-compile using Emacs 28.1:
```
ob-lesim.el:81:54: Warning: Unused lexical variable `vars'
ob-lesim.el:83:37: Warning: Unused lexical variable `result-params'
ob-lesim.el:84:46: Warning: Unused lexical variable `result-type'
In end of data:
ob-lesim.el:141:32: Warning: the function `prop-match-value' is not known to
    be defined.
ob-lesim.el:137:29: Warning: the function `prop-match-beginning' is not known
    to be defined.
ob-lesim.el:135:25: Warning: the function `text-property-search-forward' is
    not known to be defined.
```

`ob-lesim.el` with package-lint 20230605.1550:
```
3 issues found:
55:0: error: "org-babel-expand-body:lesim" doesn't start with package's prefix "ob-lesim".
55:0: error: `org-babel-expand-body:lesim' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
68:0: error: "org-babel-lesim-var-to-lesim" doesn't start with package's prefix "ob-lesim".
```

`ob-lesim.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- ob-lesim.el#L89: Mixed tab/space indentation, consider `(setq indent-tabs-mode nil)`
- ob-lesim.el#L187: Loading a package should rarely add hooks (is this an ob pattern I've forgotten about?)
```

Package and license:
- Add a LICENSE file to the repository
  See: https://github.com/licensee/licensee

drghirlanda commented:
> This branch is also including your earlier lesim-mode commits; this won't be a problem, but for future submissions you should make sure your fork-point is up to date.

Got it, thanks!

> `ob-lesim.el` with byte-compile using Emacs 28.1:
> 
> ```
> ob-lesim.el:81:54: Warning: Unused lexical variable `vars'
> ob-lesim.el:83:37: Warning: Unused lexical variable `result-params'
> ob-lesim.el:84:46: Warning: Unused lexical variable `result-type'

Fixed.

> In end of data:
> ob-lesim.el:141:32: Warning: the function `prop-match-value' is not known to
>     be defined.
> ob-lesim.el:137:29: Warning: the function `prop-match-beginning' is not known
>     to be defined.
> ob-lesim.el:135:25: Warning: the function `text-property-search-forward' is
>     not known to be defined.
> ```

I thought these functions were standard and I don't get these warnings on my machine (Emacs 28.2). The help says "Probably introduced at or before Emacs 27.1". Do you have an idea of what might be going on here?

> `ob-lesim.el` with package-lint 20230605.1550:
> 
> ```
> 3 issues found:
> 55:0: error: "org-babel-expand-body:lesim" doesn't start with package's prefix "ob-lesim".
> 55:0: error: `org-babel-expand-body:lesim' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
> 68:0: error: "org-babel-lesim-var-to-lesim" doesn't start with package's prefix "ob-lesim".
> ```

These function names are mandated by org-babel...

> 
> `ob-lesim.el` with [melpazoid](https://github.com/riscy/melpazoid):
> 
> ```
> - ob-lesim.el#L89: Mixed tab/space indentation, consider `(setq indent-tabs-mode nil)`
> - ob-lesim.el#L187: Loading a package should rarely add hooks (is this an ob pattern I've forgotten about?)
> ```

Fixed spaces/tabs issue. 

I use the hook to make it possible to run lesim-mode scripts from org-src and org-mode buffers. The TLDR is that without the hook ob-lesim would not be very useful. However, I had omitted to make the hook run only in lesim-mode org-src buffers. This is fixed now, so the hook does not do anything if you are not using lesim-mode. I have also added documentation about what the hook does so the user is aware of it.

More in detail, the Learning Simulator software that ob-lesim and lesim-mode interface with reads scripts from file. To enable running scripts from org-mode or org-src buffers, I need to create a temporary file containing a valid script, and this may require resolving noweb references from the org-mode buffer (there are specific use cases that make noweb references particularly useful  to Learning Simulator users.) Resolving noweb references is needed also for script highlighting, because lesim-mode highlights some scripting errors that can be detected only by looking at a complete script. So, in summary, the hook retrieves snippets of code from noweb references and remaps three key sequences to custom function that perform some lesim-mode-specific administration and cleanup. 

> 
> Package and license:
> 
>     * Add a LICENSE file to the repository
>       See: https://github.com/licensee/licensee

Done.

Thanks!

riscy commented:
> I thought these functions were standard and I don't get these warnings on my machine (Emacs 28.2). The help says "Probably introduced at or before Emacs 27.1". Do you have an idea of what might be going on here?

Looks like your package needs a `(require 'text-property-search)` at the top.  Most likely you're not seeing all of the compilation warnings because you're compiling your code in an instance of Emacs that has already loaded this code, or some past version of it at least.

> These function names are mandated by org-babel...

Gotcha.  I'll try to commit these to memory. :)

> without the hook ob-lesim would not be very useful

I took a better look and I see.  I think this is not a problem in this small case.

riscy commented:
I think this is small enough -- just the missing `(require ..)` mentioned above -- that I'll hit merge on this.

drghirlanda commented:
Great, thanks again for everything!

tarsius commented:
What package provides `inf-lesim`? Are you going to add that package to Melpa as well?

drghirlanda commented:
Hi! ob-lesim depends on lesim-mode. It's not an actual inf mode because the
Learning Simulator it interfaces with is non-interactive.

Happy to answer any questions!

On Sun, Jun 18, 2023, 5:57 PM Jonas Bernoulli ***@***.***>
wrote:

> What package provides inf-lesim? Are you going to add that package to
> Melpa as well?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/melpa/melpa/pull/8580#issuecomment-1596335742>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABH7EPFGNGKQ6PAUMMZLXVLXL6PX7ANCNFSM6AAAAAAY46LYSA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

tarsius commented:
`org-babel-expand-body:lesim` does `(require 'inf-lesim nil t)`. Is that a mistake?

drghirlanda commented:
Yes, it was left over from the org-babel template. Fixed it now, and thanks
for letting me know.

On Sun, Jun 18, 2023 at 9:18 PM Jonas Bernoulli ***@***.***>
wrote:

> org-babel-expand-body:lesim does (require 'inf-lesim nil t). Is that a
> mistake?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/melpa/melpa/pull/8580#issuecomment-1596346869>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABH7EPEWWX2SSITORYHA3PLXL6SGLANCNFSM6AAAAAAY46LYSA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Stefano Ghirlanda - www.intercult.su.se/~stefano - drghirlanda.com
Full Professor of Psychology, Biology, and Anthropology, Brooklyn College
Guest Professor, Stockholm University Centre for the Study of Cultural
Evolution

tarsius commented:
No problem.
